### PR TITLE
added internal fix dummy command to enable more control of fix ordering

### PR DESCRIPTION
**Pull request by sjplimp**
**Summary**

Add a fix dummy command which other styles can use to create a placeholder fix (early in the input script) which is replaced later with the desired fix.  This enables the ordering of fixes to correspond to the listing of commands in the input script.  This addresses a bug reported where granular history was not preserved correctly with fix deform or fix balance when they migrated atoms w/out FixNeighHistory having been invoked soon enough to assign history to atoms.

**Related Issues**

N/A

**Author(s)**

Steve

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Should not, but regression testing should be done.

**Implementation Notes**

_Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in LAMMPS are affected_

**Post Submission Checklist**

_Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply_

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

_Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)_

### CODE REVIEW

**Comment by sjplimp** [comment]
@akohlmey This needs to be tested by the granular person here, Joel Clemmer, to verify it fixes the issues he has been seeing.  @akohlmey @stanmoore1 This may also affect USER-OMP and KOKKOS b/c of their use of FixNeighHistory, and the 2 granular pair styles which instantiate and use it.

File: src/fix_dummy.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  initial_integrate_flag = final_integrate_flag = 0;
+  pre_exchange_flag = pre_neighbor_flag = 0;
+  pre_force_flag = post_force_flag = 0;
+  end_of_step_flag = 0;
+
+  int iarg = 3;
+  while (iarg < narg) {
+    if (strcmp(arg[iarg],"initial_integrate") == 0) initial_integrate_flag = 1;
+    else if (strcmp(arg[iarg],"final_integrate") == 0) final_integrate_flag = 1;
+    else if (strcmp(arg[iarg],"final_integrate") == 0) final_integrate_flag = 1;
```
Comment (rtoijala) [inline src/fix_dummy.cpp]: Are these lines identical on purpose? Seems like it should be initializing the other flags. Copy-paste mistake?

**Comment by akohlmey** [comment]
> @akohlmey This needs to be tested by the granular person here, Joel Clemmer, to verify it fixes the issues he has been seeing. @akohlmey @stanmoore1 This may also affect USER-OMP and KOKKOS b/c of their use of FixNeighHistory, and the 2 granular pair styles which instantiate and use it.

@sjplimp i was just looking at having the same style twice in a hybrid pair style for the neighbor list exclusions. hence my question: how is this supposed to work with the neighbor history list? doesn't each sub-style have to have its own? or should we simply disallow having those styles in hybrid multiple times?

File: src/fix_dummy.cpp (hunk 1 above)
Comment (sjplimp) [reply src/fix_dummy.cpp]: good catch - fixed those lines

**Comment by sjplimp** [comment]
Fixed a bug with indexing the replaced fix, which Joel caught in his test.

**Comment by sjplimp** [comment]
@akohlmey I think you are referring to using the generic name NEIGH_HISTORY for the pair style fix? Which could be a problem in a pair hybrid simulation.  I'll ask the granular folks here what they think. I'm guessing no one ever does that.  But we should still allow for it.  Do you like what Tim Sirk did in pair srp to generate indexed names for fix srp instances?  I presume that uses a static class variable?

**Comment by akohlmey** [comment]
> @akohlmey I think you are referring to using the generic name NEIGH_HISTORY for the pair style fix? Which could be a problem in a pair hybrid simulation. I'll ask the granular folks here what they think. I'm guessing no one ever does that. But we should still allow for it. Do you like what Tim Sirk did in pair srp to generate indexed names for fix srp instances? I presume that uses a static class variable?

It is a per-file static variable, which is effectively the same. Yes, that is a solution. It would allow for hundreds of concurrent instances with unique ids. To protect it better against race conditions (in case we ever get around to trying to make LAMMPS fully thread-safe), it would have to be done slightly different:
```
int my_srp_instance = srp_instance++;
fix_id[0] = '0' + my_srp_instance / 10;
fix_id[1] = '0' + my_srp_instance % 10;
```
But at the current state of affairs, the existing code is sufficient.

**Comment by akohlmey** [comment]
@sjplimp as far as I can tell, not changes to USER-OMP are needed. The modifications in this branch are all in functions that are inherited by the USER-OMP styles and not overridden. That seems to be different for KOKKOS, though.

**Comment by akohlmey** [comment]
@sjplimp you may want to merge in the current master, so the PR will clear the programming style check for tabs and trailing whitespace.

**Comment by sjplimp** [comment]
Joel Clemmer has done his granular testing.  So this is now good to merge,
once @stanmoore1 has checked if there are Kokkos changes needed.

**Comment by sjplimp** [comment]
@stanmoore1 Do you recall if you ever looked at this one, for any induced Kokkos changes?

**Comment by stanmoore1** [comment]
Not yet, will do.

**Comment by stanmoore1** [comment]
I'm running this through Kokkos regression testing now.

**Comment by stanmoore1** [comment]
All Kokkos regression tests are passing now for this PR.

**Pull request merged**